requirements/specifications/general/context-parameters.md

```diff
@@ -0,0 +1,303 @@
+
+All RPC methods tagged as event that have context parameters **MUST**
+have the result schema wrapped in an object. The original result schema
+**MUST** be moved to a property called data. Each of the context
+parameters **MUST** be copied to a property called context.
+
+So the following RPC event:
+
+```json
+{
+    "name": "onContextualEvent",
+    "tags": [
+        {
+            "name": "event"
+        }
+    ],
+    "params": [
+        {
+            "name": "a",
+            "required": true,
+            "schema": {
+                "type": "boolean"
+            }
+        },
+        {
+            "name": "b",
+            "required": false,
+            "schema": {
+                "type": "string"
+            }
+        },
+        {
+            "name": "c",
+            "required": false,
+            "schema": {
+                "type": "number"
+            }
+        }
+    ],
+    "result": {
+        "name": "result",
+        "schema": {
+            "type": "object",
+            "properties": {
+                "foo": {
+                    "type": "boolean"
+                }
+            }
+        }
+    }
+}
+```
+
+Would have its result transformed to:
+
+```json
+{
+    "name": "result",
+    "schema": {
+        "type": "object",
+        "properties": {
+            "data": {
+                "type": "object",
+                "properties": {
+                    "foo": {
+                        "type": "boolean"
+                    }
+                }
+            },
+            "context": {
+                "type": "object",
+                "properties": {
+                    "a": {
+                        "type": "boolean"
+                    },
+                    "b": {
+                        "type": "string"
+                    },
+                    "c": {
+                        "type": "number"
+                    }
+                },
+                "required": [
+                    "a"
+                ]
+            }
+        }
+    }
+}
+```
+
+### 4.2. JavaScript API
+
+TBD
+
+#### 4.2.1. Event Listener Signatures
+
+If any of the context parameters are optional, then a callback signature
+must be generated for each left-to-right combination of the parameters.
+
+For example, the method:
+
+```json
+{
+    "name": "onFoo",
+    "tags": [
+        {
+            "name": "event"
+        }
+    ],
+    "params": [
+        {
+            "name": "a",
+            "required": true,
+            "schema": {
+                "type": "boolean"
+            }
+        },
+        {
+            "name": "b",
+            "required": false,
+            "schema": {
+                "type": "string"
+            }
+        },
+        {
+            "name": "c",
+            "required": false,
+            "schema": {
+                "type": "number"
+            }
+        }
+    ]
+}
+```
+
+Would result in the following method signatures:
+
+```typescript
```

Comment:
Can't do a, c, because there'd be no way to know if it was a, b or a, c (if the types where the same).

Sure, in this case the types are different, but seems overly complex to generate.